troubleshooting/project-03/solution.md

```diff
@@ -0,0 +1,516 @@
+May 23 18:18:55 ip-172-31-30-191.ec2.internal systemd[1]: httpurr.service: Scheduled restart job, restart counter is at 5.
+May 23 18:18:55 ip-172-31-30-191.ec2.internal systemd[1]: Stopped httpurr.service - HTTPurr.
+May 23 18:18:55 ip-172-31-30-191.ec2.internal systemd[1]: httpurr.service: Start request repeated too quickly.
+May 23 18:18:55 ip-172-31-30-191.ec2.internal systemd[1]: httpurr.service: Failed with result 'exit-code'.
+May 23 18:18:55 ip-172-31-30-191.ec2.internal systemd[1]: Failed to start httpurr.service - HTTPurr.
+```
+
+- I looked up for more details about the server service :
+
+```sh
+[saadia@ip-172-31-30-191 httpurr-source]$ cat /etc/systemd/system/httpurr.service
+[Unit]
+Description=HTTPurr
+[Service]
+User=httpurr
+Group=httpurr
+TimeoutStartSec=0
+Restart=always
+ExecStart=/usr/local/bin/httpurr --dbengine=postgres --dbhost=ip-192-168-47-214.ec2.internal --dbport=3306 --dbuser=httpurr --dbpass=miauw --dbname=httpurr
+[Install]
+WantedBy=multi-user.target
+```
+
+- From the error message and what we have in the service file. The httpurr user don’t have access to run the server (permission denied)
+
+**Hypotheses**
+
+1. Maybe the user `httpurr`or the group `httpurr` don’t exist
+
+```sh
+[saadia@ip-172-31-30-191 httpurr-source]$ cat /etc/passwd
+root:x:0:0:root:/root:/bin/bash
+bin:x:1:1:bin:/bin:/sbin/nologin
+…
+httpurr:x:1001:1001::/home/httpurr:/bin/bash
+…
+```
+
+=> The `httpurr` user exist
+
+```sh
+[saadia@ip-172-31-30-191 httpurr-source]$ cat /etc/group
+root:x:0:
+bin:x:1:
+…
+httpurr:x:1001:
+…
+```
+
+=> The `httpurr`group exist
+
+2.  Maybe the user `httpurr` doesn't have the right permission
+
+- At the beginning I was think it’s a permission related to the file accessed by the `httpurr `user during the start of the server and checked the permission for `/usr/local/bin/httpurr` and the source code
+
+```sh
+[saadia@ip-172-31-30-191 httpurr-source]$ ls -l
+total 12
+drwxrwxr-x. 2 root root  22 May 10 09:16 catsdb
+-rw-rw-r--. 1 root root 186 May 23 15:40 go.mod
+-rw-rw-r--. 1 root root 499 May 10 09:16 go.sum
+-rw-rw-r--. 1 root root 966 May 10 09:16 httpurr.go
+drwxrwxr-x. 4 root root  61 May 23 15:39 vendor
+```
+
+=> The user `httpurr` have the right to read those files which is enough
+
+```sh
+[saadia@ip-172-31-30-191 httpurr-source]$ ls -l /usr/local/bin/httpurr
+-rwxr-xr-x. 1 root root 8508029 May 10 09:12 /usr/local/bin/httpurr
+```
+
+=> The user `httpurr` have the right to read and execute this file which is enough
+
+- I searched for the error listen tcp :80: bind: permission denied
+  And then found that 80 is a privileged port :
+
+_In Linux, a privileged port is a port in the range 1-1023. By default, privileged ports can’t be bound to non-root processes._
+
+=> So `httpurr` is a non-root user so this user can’t bind to 80
+
+- The easiest solution that comes to my mind is to bind to a different port
```

Comment:
Alternatives are:
- running it as root (this is a bad idea)
- changing the systemd config, to grant the `CAP_NET_BIND_SERVICE` capability (this is done by editing the `httpurr.service`)